### PR TITLE
Improve multi-device data loading strategy

### PR DESCRIPTION
**Pull request by laggui**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

#2277 

### Changes

The data loader strategy currently requires the batcher implementation to load the data on a specific device. For a single GPU, this is fine. But for multi-gpu, the same device (in our examples, the first) is always used and the data then has to be moved to the same device as the model before the inference step.

This is not ideal since a single GPU might get overloaded when loading the data. The `to_device` implementation for our cubecl backends goes through a CPU buffer, which makes things even worse.

This PR changes the `Batcher` trait to provide the correct device for the batch

```rust
impl<B: Backend> Batcher<B, MyItem, MyBatch<B>> for MyBatcher {
    fn batch(&self, items: Vec<MyItem>, device: &B::Device) -> MyBatch<B> {
        // The correct `device` is already provided for the impl to use
    }
}
```

To enable the device selection for data loading, the `DataLoader` trait is now generic over a backend for which a device can be set.

```rust
pub trait DataLoader<B: Backend, O>: Send {
     // [...]

    /// Sets the device for the data loader, ensuring the batches are assigned to the correct device.
    fn set_device(&mut self, device: B::Device);
    /// Returns a new data loader containing a subset of the data.
    fn slice(&self, start: usize, end: usize) -> Box<dyn DataLoader<B, O>>;
}
```

Data loaders can also be sliced to return a subset of the data. For multi-device, this allows us to split a single data loader into multiple data loaders to get one data loader per worker (i.e., one per device), so the data is loaded on the correct device before being passed to the model.

To allow the `MultiThreadDataLoader` to be split, it now holds the required fields for the dataset. The data loaders are lazily initialized when `.iter()` is called.

### Testing

Unit tests. Don't have a multi-gpu machine at hand to check the improvements 😅 though changes are fairly straightforward.

### CODE REVIEW

File: examples/guide/src/training.rs [hunk 1]
```diff
@@ -74,8 +74,8 @@ pub fn train<B: AutodiffBackend>(artifact_dir: &str, config: TrainingConfig, dev
 
     B::seed(config.seed);
 
-    let batcher_train = MnistBatcher::<B>::new(device.clone());
-    let batcher_valid = MnistBatcher::<B::InnerBackend>::new(device.clone());
+    let batcher_train = MnistBatcher::default();
+    let batcher_valid = MnistBatcher::default();
```
Comment (nathanielsimard) [inline examples/guide/src/training.rs:78]: It could be the same batcher now.

Comment (laggui) [reply examples/guide/src/training.rs:78]: Yupp! Even simpler 😄

File: crates/burn-core/src/data/dataloader/lazy.rs [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+    ///
+    /// # Returns
+    /// A vector of lazy data loaders, each initialized with one of the provided distribution strategies.
+    /// Each returned data loader represents a portion of the original dataset and can be independently
+    /// processed. The number of returned data loaders matches the number of distribution strategies provided.
+    fn split(
+        &self,
+        distributors: Vec<Box<dyn DistributionStrategy<Resource = Self::Resource>>>,
+    ) -> Vec<Box<dyn LazyDataLoader<O, Resource = Self::Resource>>>;
+}
```
[the remark this answers is not in the document]
Comment (laggui) [reply crates/burn-core/src/data/dataloader/lazy.rs]: Hmmm not sure how this would help. This trait is used to assign different dispatchers to each data loader split. The dispatcher is linked to the data loader, not dataset.

Essentially this trait was added to split a data loader into multiple data loaders, each with a partial view of the dataset (so `PartialDataset` is actually used in the implementation) and a specific dispatcher.

So for multi-GPU training, the original data loader is created for the entire dataset. But when multiple devices are specified, the data loader is split into `num_devices` partial data loaders that dispatch a batch for a specific device. That way, each worker is responsible for a single data loader and device.

The trait could be removed, but it makes it more difficult since we're always working with a `dyn DataLoader` that is created by the user for the learner.

Lmk what you think.

[the remark this answers is not in the document]
Comment (laggui) [reply crates/burn-core/src/data/dataloader/lazy.rs]: This has been further simplified. Data loaders are now generic over the backend to allow device selection/assignment and can be sliced.

File: crates/burn-core/src/data/dataloader/builder.rs [hunk 3]
```diff
@@ -91,27 +112,137 @@ where
     /// # Returns
     ///
     /// The data loader.
-    pub fn build<D>(self, dataset: D) -> Arc<dyn DataLoader<O>>
+    pub fn build<D>(self, dataset: D) -> Arc<dyn DataLoader<B, O>>
```
Comment (nathanielsimard) [inline crates/burn-core/src/data/dataloader/builder.rs:115]: If we put the `DataLoader` trait inside an Arc, we should remove the `DynDataLoader` trait.

File: crates/burn-core/src/data/dataloader/multithread.rs [hunk 4]
```diff
@@ -26,32 +46,94 @@ struct MultiThreadsDataloaderIterator<O> {
...
+            batcher,
+            num_threads,
+            device,
+            rng,
+            dataloaders: OnceCell::new(),
+        }
+    }
+
+    /// Force initialization if needed.
+    fn initialize(&self) -> &[Box<dyn DynDataLoader<B, O>>] {
```
Comment (nathanielsimard) [inline crates/burn-core/src/data/dataloader/multithread.rs:89]: Could we use `Arc<dyn DataLoader<B, O>` here instead?

File: crates/burn-train/src/learner/train_val.rs [hunk 5]
```diff
@@ -144,14 +144,40 @@ impl<LC: LearnerComponents> Learner<LC> {
...
+                };
+                let mut dataloader = dataloader_train.slice(start, end);
+                dataloader.set_device(self.devices[i].clone());
+                dataloaders.push(Arc::from(dataloader));
+                start = end;
+            }
+            dataloaders
+        } else {
+            vec![dataloader_train]
+        };
```
Comment (nathanielsimard) [inline crates/burn-train/src/learner/train_val.rs:171]: I would extract that into a new function, maybe called: `split_dataloaders(dataloader, &devices)`

File: crates/burn-train/src/learner/train_val.rs [hunk 6]
```diff
@@ -175,6 +201,7 @@ impl<LC: LearnerComponents> Learner<LC> {
                 break;
             }
 
+            // TODO: multi-device validation?
```
Comment (nathanielsimard) [inline crates/burn-train/src/learner/train_val.rs:204]: Still relavant?

Comment (laggui) [reply crates/burn-train/src/learner/train_val.rs:204]: Forgot that I added this comment 😅 was mostly for my own sake

The comment is still relevant I guess, cuz the validation loop still runs on a single device (same behavior as before). We just split the training across GPUs.

For most data parallel use cases this is fine I think, unless the validation set is huge splitting across multiple GPUs probably doesn't have as much of a benefit. But we could do it in the future.

File: crates/burn-core/src/data/dataloader/multithread.rs (hunk 4 above)
Comment (laggui) [reply crates/burn-core/src/data/dataloader/multithread.rs:89]: Yeah probably, but in the implementation we share each data loader to a thread.

With a box and `clone_dyn()` this didn't require any further requirements for `Sync`. So if we switch to `Arc<dyn DataLoader<B, O>>` we probably need to have another bound (and propagate it).

File: crates/burn-core/src/data/dataloader/builder.rs (hunk 3 above)
Comment (laggui) [reply crates/burn-core/src/data/dataloader/builder.rs:115]: Yeah that makes sense. But I'm curious why this wasn't already the case?

File: crates/burn-core/src/data/dataloader/multithread.rs (hunk 4 above)
Comment (laggui) [reply crates/burn-core/src/data/dataloader/multithread.rs]: Updated:

`MultiThreadDataLoader` simply keeps a vec of `BatchDataLoader`. The `Box<dyn DataLoader>` was superfluous, left over from the previous impl. `BatchDataLoader` is always used to create an instance.

```rust
pub struct MultiThreadDataLoader<B: Backend, I, O> {
    // Configuration parameters needed for initialization
    strategy: Box<dyn BatchStrategy<I>>,
    dataset: Arc<dyn Dataset<I>>,
    batcher: Box<dyn DynBatcher<B, I, O>>,
    device: B::Device,
    rng: Option<rand::rngs::StdRng>,
    num_threads: usize,

    // The lazily initialized data loaders
    dataloaders: OnceCell<Vec<BatchDataLoader<B, I, O>>>,
}
```

so `initialize()` now returns a slice of `BatchDataLoader`s, and we can remove the `DynDataLoader` trait with `clone_dyn()`.
